### PR TITLE
Add demo to run Rust natively in a WASM VM

### LINKED ISSUE #16: WASM: Use wasmtime-embed to implement callable Rust functions in Wasmtime
A longer-term WASM priority for Enarx is the ability to implement custom WASI functions in Wasmtime (as detailed in #10). There are a number of approaches to achieve that goal; this issue focuses on one particular way.

Yury from Mozilla has recently been working on tools to make Wasmtime [embeddable](https://github.com/yurydelendik/wasmtime-embed). While they have a number of uses, one particularly interesting result from his work is the ability to write a native Rust function and have it included in the set of functions that Wasmtime imports, and have it usable by WASM binaries. We'd like to investigate this technique as a possibility to achieve custom WASI implementations.

### PR DESCRIPTION
**Pull request by steveej**
Fixes #16.

### CODE REVIEW

**Review by mbestavros — COMMENTED** [review]
Looks good to me -- I don't notice significant changes from what you showed us offline last Friday, and we took a pretty deep dive into what everything does and why it needs to be there at that point.

If I were to nitpick, a few inline comments to explain what exactly `main.rs` is doing would be helpful for people looking at this code later.

**Comment by steveej** [comment]
Thanks @mbestavros, I added a few comments and also changed the variable names to be more descriptive.

**Comment by mbestavros** [comment]
Thanks -- definitely more immediately understandable now.

I'll defer to Nathaniel for final merge approval, but this looks good to me as-is.

File: wasmtime-native-embed/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    /// This trait defines an example interface for the implementations which we
+    /// want to import to WASM.
+    #[wasm_import]
+    pub trait Calc {
+        fn add(&self, a: u32, b: u32) -> u32;
+        fn sub(&self, a: u32, b: u32) -> u32;
+        fn mul(&self, a: u32, b: u32) -> u32;
+    }
+
+    pub struct PrintingCalc {}
```
Comment (npmccallum) [inline wasmtime-native-embed/src/main.rs]: I think idiomatically we want to use a unit struct here. So: `pub struct PrintingCalc();`.

File: wasmtime-native-embed/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+        fn mul(&self, a: u32, b: u32) -> u32 {
+            let result = a * b;
+            print!("{} * {}", a, b);
+            result
+        }
+    }
+}
+
+use app::{PrintingCalc, Calc};
```
Comment (npmccallum) [inline wasmtime-native-embed/src/main.rs]: Nit pick: this is probably over-verbose unless you're trying to restrict line length. `app::PrintingCalc` is very readable.

File: wasmtime-native-embed/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+fn main() {
+    // We need an instance of the implementing struct to import it to WASM
+    let printing_calc = PrintingCalc {};
```
Comment (npmccallum) [inline wasmtime-native-embed/src/main.rs]: I don't mind long variable names. But I think Rust idiomatically prefers short ones. Does Rust give any official guidance on this? For example: is `pcalc` or `printing_calc` preferred? I'm honestly not sure. But it would be good to know.

File: wasmtime-native-embed/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+name = "wasmtime-native-embed"
+version = "0.1.0"
+authors = ["Stefan Junker <sjunker@redhat.com>"]
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+wasmtime-embed = { git = "https://github.com/yurydelendik/wasmtime-embed", package = "wasmtime-embed", rev = "cc5740f" }
+wasmtime-embed-macro = { git = "https://github.com/yurydelendik/wasmtime-embed", package = "wasmtime-embed-macro", rev = "cc5740f" }
```
Comment (npmccallum) [inline wasmtime-native-embed/Cargo.toml:13]: Since we are obviously using a git repo based dependency, it would be good to explain why. It would also be good to explain what the path is towards making this a standard `crates.io` dependency is.

**Review by npmccallum — CHANGES_REQUESTED** [review]
This overall looks very good to me. I've requested at least one minor change and some documentation of the `wasmtime-embed` situation.

File: wasmtime-native-embed/src/main.rs (hunk 2 above)
Comment (steveej) [reply wasmtime-native-embed/src/main.rs]: Do you suggest to split this into two separate lines?

Comment (npmccallum) [reply wasmtime-native-embed/src/main.rs]: No. I mean remove the `use` statement altogether.

I generally try to balance my use of namespace importing with concerns like readability, line length, etc. If you think this is worthwhile, then maybe move it within the `main()` function?

File: wasmtime-native-embed/src/main.rs (hunk 3 above)
Comment (steveej) [reply wasmtime-native-embed/src/main.rs]: I find long variable names very helpful and generally use them. Of course it is not the length which is significant, but the variable name as an indicator of what it's used for, in the style of the [Hungarian Notation](https://en.wikipedia.org/wiki/Hungarian_notation).

I haven't looked at any guidelines for this but I find the short variables names found in the C ecosystem extremely confusing, which usually puts a barrier between me and a project I'd like to possibly contribute.

File: wasmtime-native-embed/Cargo.toml (hunk 4 above)
Comment (steveej) [reply wasmtime-native-embed/Cargo.toml:13]: I haven't thought about this much. I don't expect that specific project to ever make a release, as the macros, or a variation thereof, will likely be integrated with wasmtime-api or a different wasmtime subcrate at one point.

File: wasmtime-native-embed/src/main.rs (hunk 3 above)
Comment (npmccallum) [reply wasmtime-native-embed/src/main.rs]: I don't have a strong opinion here. (Except I detest embedding the type of a variable in its name...) But I would like to have an Enarx policy. I'm not requesting any change on this PR. I just want us to start a conversation.

File: wasmtime-native-embed/Cargo.toml (hunk 4 above)
Comment (npmccallum) [reply wasmtime-native-embed/Cargo.toml:13]: Just an inline comment saying basically the above would be fine.

File: wasmtime-native-embed/src/main.rs (hunk 3 above)
Comment (mbestavros) [reply wasmtime-native-embed/src/main.rs]: I agree with Stefan; I'm personally in favor of descriptive (-enough) variable names. Abbreviations can make things hard to parse.

Comment (steveej) [reply wasmtime-native-embed/src/main.rs]: I very rarely embed the type in it the variable name. An exception might be if the scope is just doing conversion between types and I don't want or can't declare it as a type in the language itself, although I do prefer the latter.
I usually give a functional description of what the variable is used for. 

I think the Hungarian Notation is a good fit which we could set as a baseline with the explicit deviation of not including the type in the name, unless it's the main logic of the scope.

**Comment by steveej** [comment]
@npmccallum thanks for the review. I acted on all comments except for the naming convention one. That should go in the top-level contribution guidelines.

File: wasmtime-native-embed/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    /// This trait defines an example interface for the implementations which we
+    /// want to import to WASM.
+    #[wasm_import]
+    pub trait Calc {
+        fn add(&self, a: u32, b: u32) -> u32;
+        fn sub(&self, a: u32, b: u32) -> u32;
+        fn mul(&self, a: u32, b: u32) -> u32;
+    }
+
+    pub struct PrintingCalc();
```
Comment (npmccallum) [inline wasmtime-native-embed/src/main.rs]: I'm really sorry. I totally told you the wrong thing before. For a unit struct, the idiom is `pub struct PrintincCalc;`.

File: wasmtime-native-embed/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+            result
+        }
+    }
+}
+
+fn main() {
+    use app::{Calc, PrintingCalc};
+
+    // We need an instance of the implementing struct to import it to WASM
+    let printing_calc = PrintingCalc {};
```
Comment (npmccallum) [inline wasmtime-native-embed/src/main.rs]:
```rust
let printing_calc = PrintingCalc {};
```

I'm really surprised that still compiles... Regardless, it needs to be updated to:

```rust
let printing_calc = PrintingCalc;
```

**Review by npmccallum — CHANGES_REQUESTED** [review]

File: wasmtime-native-embed/src/main.rs (hunk 5 above)
Comment (steveej) [reply wasmtime-native-embed/src/main.rs]: no worries at all

**Review by npmccallum — APPROVED** [review]
